File: 2018-12/2018-12-22/LockableBinaryTree.cpp
#include <iostream>

struct Node
{
public:
	int value;

	Node* left;
	Node* right;
	Node* parent;

	bool isLocked;
	unsigned int descendantLockCount;

	Node(int value, Node* left = nullptr, Node* right = nullptr, Node* parent = nullptr) noexcept
		: value(value), left(left), right(right), parent(parent), isLocked(false), descendantLockCount(0)
	{
		if (left != nullptr)
		{
			left->parent = this;
		}

		if (right != nullptr)
		{
			right->parent = this;
		}
	}

	~Node() noexcept
	{
		if (left != nullptr)
		{
			delete left;
		}

		if (right != nullptr)
		{
			delete right;
		}
	}

	inline bool IsLocked() const { return isLocked; }

	bool Lock()
	{
		if (CanBeLocked())
		{
			isLocked = true;

			Node* currentNode = parent;

			while (currentNode != nullptr)
			{
				currentNode->descendantLockCount++;
				currentNode = currentNode->parent;
			}

			return true;
		}
		
		return false;
	}

	bool Unlock()
	{
		if (CanBeLocked())
		{
			isLocked = false;

			Node* currentNode = parent;

			while (currentNode != nullptr)
			{
				currentNode->descendantLockCount--;
				currentNode = currentNode->parent;
			}

			return true;
		}

		return false;
	}

private:
	bool CanBeLocked()
	{
		if (descendantLockCount > 0)
		{
			return false;
		}

		Node* currentNode = parent;

		while (currentNode != nullptr)
		{
			if (currentNode->isLocked)
			{
				return false;
			}

			currentNode = currentNode->parent;
		}

		return true;
	}
};

int main(int argc, const char* argv[])
{
	Node* leftleft = new Node(67);
	Node* leftright = new Node(49);
	
	Node* rightright = new Node(36);

	Node* left = new Node(82, leftleft, leftright);
	Node* right = new Node(49, nullptr, rightright);

	Node* root = new Node(19, left, right);

	std::cout << std::boolalpha;

	std::cout << left->Lock() << "\n";
	std::cout << leftright->Lock() << "\n";

	left->Unlock();

	std::cout << leftright->Lock() << "\n";

	if (leftright->IsLocked())
	{
		std::cout << "Node leftright is locked." << std::endl;
	}

	std::cin.get();

	return 0;
}

File: Days 21 - 30/Day 24/LockableBinaryTree.cs
using System;

namespace DailyCodingProblem
{
	internal class Node
	{
		public int Value { get; set; }
		public bool IsLocked { get; private set; } = false;

		private Node left;
		private Node right;
		private Node parent = null;

		private int descendantLockCount = 0;

		public Node(int value, Node left = null, Node right = null)
		{
			this.Value = value;

			this.left = left;
			this.right = right;

			if (left != null)
			{
				left.parent = this;
			}

			if (right != null)
			{
				right.parent = this;
			}
		}

		public bool Lock()
		{
			if (CanBeLocked())
			{
				IsLocked = true;
				Node currentNode = parent;

				while (currentNode != null)
				{
					currentNode.descendantLockCount++;
					currentNode = currentNode.parent;
				}

				return true;
			}

			return false;
		}

		public bool Unlock()
		{
			if (CanBeLocked())
			{
				IsLocked = false;
				Node currentNode = parent;

				while (currentNode != null)
				{
					currentNode.descendantLockCount--;
					currentNode = currentNode.parent;
				}

				return true;
			}

			return false;
		}

		private bool CanBeLocked()
		{
			if (descendantLockCount > 0)
			{
				return false;
			}

			Node currentNode = parent;

			while (currentNode != null)
			{
				if (currentNode.IsLocked)
				{
					return false;
				}

				currentNode = currentNode.parent;
			}

			return true;
		}
	}

	internal class Day24
	{
		private static int Main(string[] args)
		{
			Node leftLeft = new Node(67);
			Node leftRight = new Node(43);
			Node rightRight = new Node(36);
			Node left = new Node(82, leftLeft, leftRight);
			Node right = new Node(49, null, rightRight);
			Node root = new Node(19, left, right);

			Console.WriteLine(left.Lock());
			Console.WriteLine(leftRight.Lock());

			left.Unlock();

			Console.WriteLine(leftRight.Lock());

			if (leftRight.IsLocked)
			{
				Console.WriteLine("Node leftRight is locked.");
			}

			Console.ReadLine();

			return 0;
		}
	}
}